// @mikro-orm/cli/commands/GenerateCacheCommand.js
import { MetadataDiscovery, MetadataStorage, colors } from '@mikro-orm/core';
import { FileCacheAdapter } from '@mikro-orm/core/fs-utils';
import { CLIHelper } from '../CLIHelper.js';
export class GenerateCacheCommand {
  command = 'cache:generate';
  describe = 'Generate metadata cache';
  builder = args => {
    args.option('ts', {
      type: 'boolean',
      desc: `Generate development cache for '.ts' files`,
    });
    args.option('combined', {
      alias: 'c',
      desc: `Generate production cache into a single JSON file that can be used with the GeneratedCacheAdapter.`,
    });
    return args;
  };
  /**
   * @inheritDoc
   */
  async handler(args) {
    const options = args.combined ? { combined: './metadata.json' } : {};
    const config = await CLIHelper.getConfiguration(args.contextName, args.config, {
      metadataCache: { enabled: true, adapter: FileCacheAdapter, options },
    });
    await config.getMetadataCacheAdapter().clear();
    config.set('logger', CLIHelper.dump.bind(null));
    config.set('debug', true);
    const discovery = new MetadataDiscovery(new MetadataStorage(), config.getDriver().getPlatform(), config);
    await discovery.discover(args.ts ?? false);
    const combined = args.combined && config.get('metadataCache').combined;
    CLIHelper.dump(
      colors.green(
        `${combined ? 'Combined ' : ''}${args.ts ? 'TS' : 'JS'} metadata cache was successfully generated${combined ? ' to ' + combined : ''}`,
      ),
    );
  }
}
